mlflow/genai/agent_server/server.py

```diff
@@ -0,0 +1,231 @@
+STREAM_KEY = "stream"
+
+AgentType = Literal["ResponsesAgent"]
+
+
+@experimental(version="3.6.0")
+class AgentServer:
+    """FastAPI-based server for hosting agents.
+
+    Args:
+        agent_type: An optional parameter to specify the type of agent to serve. If provided,
+        input/output validation and streaming tracing aggregation will be done automatically.
+
+        Currently only "ResponsesAgent" is supported.
+
+        If ``None``, no input/output validation and streaming tracing aggregation will be done.
+        Default to ``None``.
+    """
+
+    def __init__(self, agent_type: AgentType | None = None):
+        self.agent_type = agent_type
+        if agent_type == "ResponsesAgent":
+            self.validator = ResponsesAgentValidator()
+        else:
+            self.validator = BaseAgentValidator()
+
+        self.app = FastAPI(title="Agent Server")
+        self._setup_routes()
+
+    @staticmethod
+    def _get_in_memory_trace(trace_id: str) -> dict[str, Any]:
+        with InMemoryTraceManager.get_instance().get_trace(trace_id) as trace:
+            return {"trace": trace.to_mlflow_trace().to_dict()}
+
+    def _setup_routes(self) -> None:
+        @self.app.post("/invocations")
+        async def invocations_endpoint(request: Request):
+            # Capture headers such as x-forwarded-access-token
+            # https://docs.databricks.com/aws/en/dev-tools/databricks-apps/auth?language=Streamlit#retrieve-user-authorization-credentials
+            set_request_headers(dict(request.headers))
+
+            try:
+                data = await request.json()
+            except Exception as e:
+                raise HTTPException(status_code=400, detail=f"Invalid JSON in request body: {e!s}")
+
+            logger.debug(
+                "Request received",
+                extra={
+                    "agent_type": self.agent_type,
+                    "request_size": len(json.dumps(data)),
+                    "stream_requested": data.get(STREAM_KEY, False),
+                },
+            )
+
+            is_streaming = data.get(STREAM_KEY, False)
+            request_data = {k: v for k, v in data.items() if k != STREAM_KEY}
+
+            try:
+                request_data = self.validator.validate_and_convert_request(request_data)
+            except ValueError as e:
+                raise HTTPException(
+                    status_code=400,
+                    detail=f"Invalid parameters for {self.agent_type}: {e}",
+                )
+
+            if is_streaming:
+                return await self._handle_stream_request(request_data)
+            else:
+                return await self._handle_invoke_request(request_data)
+
+        @self.app.get("/health")
+        async def health_check() -> dict[str, str]:
+            return {"status": "healthy"}
+
+    async def _handle_invoke_request(self, request: dict[str, Any]) -> dict[str, Any]:
+        from mlflow.genai.agent_server import _invoke_function
```

Comment:
Since this is due to circular import, you may want to move _invoke_function into this file as well?